### ISSUE TITLE
Response body can't be a hash

### ISSUE THREAD

**Issue by Morred** [bblimke/webmock#449]
When I stub a request and specify a hash for the response body, it gives me the following error:
```
WebMock::Response::InvalidBody:
    must be one of: [Proc, IO, Pathname, String, Array]. 'Hash' given
```
Is there a particular reason why it can't be a hash?

**Comment by bblimke** [comment]
Please have a look at https://github.com/bblimke/webmock/pull/427

If response body is declared as a hash, it's not clear what should the hash be encoded to. JSON, XML, or perhaps url params? Therefore WebMock expects you to stringify hash to expected encoding. I.e {a: b}.to_json

**Comment by Morred** [comment]
(Sorry for the late response, and thanks for your reply.)

I see, that makes sense. Maybe this should be mentioned in the Readme?

**Comment by joshuaswilcox** [comment]
I am not sure that makes sense.  If a server is actually returning JSON as the body content, the code parsing can assume its a Hash content type, therefore calling JSON.parse(response) will fail if its already a Hash.  So its seems to me that Webmock should be able to send that as the response body, especially when a content-type of 'application/json' is explicitly passed.

**Comment by bblimke** [comment]
"If a server is actually returning JSON as the body content" - which server do you mean? We are returning a stubbed response. This response will be passed through specific http client library internals. You can't expect these libs to handle Hash.

WebMock would have to convert Hash declared in `to_return` to json string automatically first, but how would it know whether to convert this Hash to json or to xml?

**Comment by joshuaswilcox** [comment]
Would it not be best to use the headers content-type to determine whether to return JSON or XML? I am stubbing as so

    stub_request(:post,'http://......').
         to_return(
          :status => 200,
          :body => '{"transaction_status":"declined"}',
          :headers => {"Content-Type"=> "application/json"})

but would prefer not to have to call JSON.parse(response_body) since that is not a step that is required in my code as JSON is implicitly declared.  Maybe I am missing something though?  FWIW, I am using Faraday to make the actual request.

**Comment by bblimke** [comment]
In case there is response content type declared it could work. Webmock would have to raise an error is case body is declared as a hash and headers are not declared.

**Comment by bblimke** [comment]
Most people will forget about headers.

**Comment by joshuaswilcox** [comment]
I don't think that is a problem to require a content-type when passing certain types of body content, since that is generally a good practice across http requests.  Could i make a pull request for this functionality?

**Comment by bblimke** [comment]
I don't think it's a problem either, as long as user gets a clear error message explaining content type needs to be added if missing.
Pull request would be most appreciated :)

**Comment by joshuaswilcox** [comment]
Great, thanks!

**Comment by s2t2** [comment]
I support and encourage implementation of this feature.

I'm surprised this isn't a more widespread issue. I've already encountered this issue on two separate occasions with two different web services. Most recently I needed to implement a work-around whereby I changed my response-parsing method to handle/convert the stubbed response strings when in reality outside of the test environment this extra handling would not be required.

**Comment by BKDruby** [comment]
How we could add headers with `.to_rack` method?

**Comment by pboling** [comment]
I am running into this now.  How is one supposed to mock a JSON response where the response is a hash?  Not a JSON encoded string, but a hash.  At the Ruby later the response is a hash in my real code, so it must also be in my test.

**Comment by bblimke** [comment]
@pboling 

HTTP response body is never a hash. Your ruby HTTP client most likely does the response parsing for you and converts is to a hash. In order for your http client to know that the response body is JSON, you need to indicate that by the Content-Type header.

````
stub_request(:get, 'http://......').
     to_return(
      body: json_string,
      headers: {"Content-Type"=> "application/json"})
````

**Comment by dduqueti** [comment]
Thank you all for your answers, specially @joshuaswilcox and @bblimke. I was having issues with a spec for hours which mixed `WebMock` response with another gem expecting a hash. I wasn't until I ended up here that I was able to worked through.  Had to use both `JSON.generate` and `Content-Type` option.

I'm leaving my solution for other readers:

`
stub_request(:get, /exampleurl.com/)
  .to_return(
          status: 200,
          body: JSON.generate(response_json),
          headers: {"Content-Type"=> "application/json"}
        )
`

**Comment by brian-lutz-informed** [comment]
Thank you so much @bblimke and @joshuaswilcox @dduqueti.

**Comment by jeffdill2** [comment]
I think part of what makes this particularly confusing is that if you pass an empty hash, it doesn't throw an error.
